feat: add support for v7 transports

guilhermelimak opened the pull request:
This PR will add support for node-tee to be used as a pino v7 transport.

The code is still a bit crude so any suggestions are welcome

Closes #90.

guilhermelimak commented:
@mcollina I kept the original behavior intact for now and the new transport can be accessed as 'pino-tee/v7'. Once the default export on the index file is changed it will be a breaking change for anyone that's using the tee function directly.
What should I do here? Should I change the default export and bump the major version or should we maintain backwards compatibility?

mcollina commented:
CI is failing

simoneb commented:
@guilhermelimak this will be a breaking change so it's fine to bump the major. Please fix CI!

guilhermelimak commented:
There's apparently a bug on nyc (or istanbul) that prevents some lines to be marked as covered on windows. On OSX and linux everything seems to be working fine.

The lines marked as not covered on windows are confirmed to be running during the tests but for some reason they're not included on the test coverage.

This may be a bug on nyc and it was present before the node-tap update or v7 transports update (as seen [here](https://github.com/pinojs/pino-tee/runs/5404633881?check_suite_focus=true#step:5:127)). It's only getting attention now because the new node-tap versions enforce 100% coverage by default, which was making the windows CI tests to fail.

For now I deactivated the 100% coverage check and I think that if needed we should open a separate issue for trying to investigate this. I tested it on my windows machine and couldn't find out exactly what was causing the bug or any solutions/workarounds.

This PR should be ready for review now.

guilhermelimak commented:
I've improved the PR a bit more, update the exports, added an example and updated the README.

The only issue now is pino v7 won't work with node v10 which is why the v10 builds are failing now. I guess we could remove them and drop support to node v10 since pino itself won't support it anymore?

cc @simoneb @mcollina 

This is the error I get when trying to use pino v7 with node v10.24:


```
Error: Cannot find module 'worker_threads'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
    at Function.Module._load (internal/modules/cjs/loader.js:562:25)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at require (internal/modules/cjs/helpers.js:25:18)
    at Object.<anonymous> (/home/xx/proj/pino-tee/node_modules/pino/lib/tools.js:29:26)
```

simoneb commented:
@guilhermelimak let's drop support for Node 10 in this PR as pino also dropped support. worker_threads are now a required feature

guilhermelimak commented:
@simoneb done

simoneb commented on tee.js:
```diff
@@ -61,8 +64,7 @@ function start () {
   if (args._.length % 2) {
-    console.error('pino-tee requires an even amount of args')
-    console.error('Usage: pino-tee [filter dest]..')
+    console.error('pino-tee requires an even amount of args\nUsage: pino-tee [filter dest].')
```
```suggestion
    console.error('pino-tee requires an even number of args\nUsage: pino-tee [filter dest].')
```

simoneb commented on test/bin.js:
```diff
@@ -86,8 +86,7 @@ test('invoked with incorrect args', (t) => {
     t.same(arr, [
-      'pino-tee requires an even amount of args\n',
-      'Usage: pino-tee [filter dest]..\n'
+      'pino-tee requires an even amount of args\nUsage: pino-tee [filter dest].\n'
```
```suggestion
      'pino-tee requires an even number of args\nUsage: pino-tee [filter dest].\n'
```

simoneb commented on package.json:
```diff
@@ -37,16 +37,17 @@
     "fast-json-parse": "^1.0.3",
     "minimist": "^1.2.5",
-    "pino": "^6.9.0",
+    "pino": "^7.10.0",
```
I wonder why  this is not a peer dependency instead

simoneb commented on tee.js:
```diff
@@ -49,7 +50,9 @@ function alwaysTrue () {
 }
 
-module.exports = tee
+module.exports = transport
+
+module.exports.tee = tee
```
I suspect this change is because the transport mechanism requires it to be the main export, right? this also means that this is a breaking change for Node.js users. Which I guess is fine too

guilhermelimak replied:
yep, exactly. I did change the docs to reflect this change.

simoneb replied:
can I ask you do to do one thing? I'm ok with this change but for consistency can we check how other pino transports have handled this when introducing support for v7? I suspect they had to do something similar and we should make sure we do it in the same way

guilhermelimak replied:
loks like pino pretty did basically the same thing [here](https://github.com/pinojs/pino-pretty/commit/72808da44b708bbc3b50fbb3ecddb89f706a458c#diff-e727e4bdf3657fd1d798edcd6b099d6e092f8573cba266154583a746bba0f346R209)

simoneb reviewed (approved):
overall LGTM, a few comments

simoneb reviewed (approved):
Perfect, LGTM

Fdawgs commented on .taprc:
```diff
@@ -0,0 +1 @@
+check-coverage: false
```
Disabling 100% coverage checking?

mcollina replied:
please restore this

guilhermelimak replied:
As mentioned on https://github.com/pinojs/pino-tee/pull/94#issuecomment-1109080621, this is keeping the old behavior (on tap@14 100% coverage enforcement is disabled by default). The nyc bug on windows will prevent us to enable it since it would break our CI.

mcollina reviewed (changes requested):
restore 100% coverage

simoneb commented:
@mcollina @Fdawgs from https://github.com/pinojs/pino-tee/pull/94#issuecomment-1109080621

> For now I deactivated the 100% coverage check and I think that if needed we should open a separate issue for trying to investigate this. I tested it on my windows machine and couldn't find out exactly what was causing the bug or any solutions/workarounds.

Fdawgs commented:
> @mcollina @Fdawgs from [#94 (comment)](https://github.com/pinojs/pino-tee/pull/94#issuecomment-1109080621)

Cheers @simoneb, missed that.

simoneb commented:
@guilhermelimak let's follow up right away with an issue just to avoid any chance of forgetting to look into this please

mcollina reviewed (approved):
lgtm
